### PR TITLE
avm2: Improve XML set_property_local implementation

### LINKED ISSUE #12162: Cannot set an XML/XMLList object as an element yet
At the moment it is not possible to set an XML/XMLList object as an element on another XMLList object.

Is there a possibility that it will be implemented soon or is it not a priority at the moment? Would that be a lot of effort?

Thanks for any answers in advance!

### LINKED ISSUE #12910: Laser Cannon 2 can't push to start....
### Describe the bug

[laser-cannon-2.zip](https://github.com/ruffle-rs/ruffle/files/12415114/laser-cannon-2.zip)

# Player Info
Allows script access: false
Renderer: wgpu
Adapter Backend: Gl
Adapter Name: "ANGLE (Intel, Intel(R) HD Graphics 630 Direct3D11 vs_5_0 ps_5_0, D3D11)"
Adapter Device Type: IntegratedGpu
Adapter Driver Name: ""
Adapter Driver Info: ""
Enabled features: Features(TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES | PUSH_CONSTANTS)
Available features: Features(TEXTURE_COMPRESSION_BC | CLEAR_TEXTURE | MULTIVIEW)
Current limits: Limits { max_texture_dimension_1d: 16384, max_texture_dimension_2d: 16384, max_texture_dimension_3d: 2048, max_texture_array_layers: 256, max_bind_groups: 4, max_bindings_per_bind_group: 640, max_dynamic_uniform_buffers_per_pipeline_layout: 8, max_dynamic_storage_buffers_per_pipeline_layout: 0, max_sampled_textures_per_shader_stage: 16, max_samplers_per_shader_stage: 16, max_storage_buffers_per_shader_stage: 0, max_storage_textures_per_shader_stage: 0, max_uniform_buffers_per_shader_stage: 11, max_uniform_buffer_binding_size: 16384, max_storage_buffer_binding_size: 0, max_vertex_buffers: 8, max_buffer_size: 268435456, max_vertex_attributes: 16, max_vertex_buffer_array_stride: 255, min_uniform_buffer_offset_alignment: 256, min_storage_buffer_offset_alignment: 256, max_inter_stage_shader_components: 60, max_compute_workgroup_storage_size: 0, max_compute_invocations_per_workgroup: 0, max_compute_workgroup_size_x: 0, max_compute_workgroup_size_y: 0, max_compute_workgroup_size_z: 0, max_compute_workgroups_per_dimension: 0, max_push_constant_size: 96 }
Surface quality: high
Surface samples: 4
Surface size: Extent3d { width: 550, height: 400, depth_or_array_layers: 1 }

# Page Info
Page URL: https://ruffle-regressions.github.io/

# Browser Info
User Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36
Platform: Win32
Has touch support: false

# Ruffle Info
Version: 0.1.0
Name: nightly 2023-08-23
Channel: nightly
Built: 2023-08-23T00:19:38.921Z
Commit: d55cc23f5503f53932c64edf516ca375eaa01764
Is extension: false

# Metadata
width: 700
height: 525
frameRate: 40
numFrames: 2
swfVersion: 10
backgroundColor: undefined
isActionScript3: true
uncompressedLength: 3856913

### Expected behavior

Laser Cannon 2 can't push to start....

### Affected platform

Self-hosted version

### Operating system

Windows 10

### Browser

Google Chrome 115.0.5790.110 [64 bits]

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by sleepycatcoding**
~~This is a draft as it currently makes test `xml_namespaced_property` fail.~~
~~Also requires #12993~~ (merged)

Fixes #12162
Fixes #12910

### CODE REVIEW

File: core/src/avm2/e4x.rs [hunk 1]
```diff
@@ -244,6 +244,57 @@ impl<'gc> E4XNode<'gc> {
         node
     }
 
+    /// Returns the amount of children in this node if this node is of Element kind, otherwise returns [None].
+    pub fn length(&self) -> Option<usize> {
+        let E4XNodeKind::Element { children, .. } = &*self.kind() else {
```
Comment (adrian17) [inline core/src/avm2/e4x.rs:249]: nit: I'd say do it the other way around?
```
it let E4XNodeKind::Element { children, .. } = &*self.kind() {
    Some(children.len())
} else {
    None
}
```

File: core/src/avm2/object/xml_object.rs [hunk 2]
```diff
@@ -328,50 +360,84 @@ impl<'gc> TObject<'gc> for XmlObject<'gc> {
             return Ok(());
         }
 
-        let self_node = *self.node();
-        let write = self.0.write(mc);
-        let mut kind = write.node.kind_mut(mc);
-        let E4XNodeKind::Element { children, .. } = &mut *kind else {
+        // 7. Let isValidName be the result of calling the function isXMLName (section 13.1.2.1) with argument n
+        let is_valid_name = name.local_name().map_or(Ok(false), |x| {
+            crate::avm2::e4x::is_xml_name(activation, x.into())
```
Comment (adrian17) [inline core/src/avm2/object/xml_object.rs:365]: (in retrospect, maybe passing an AvmString to `is_xml_name` would have been a nicer design, so it wouldn't need a `Result` here)

**Review by adrian17 — APPROVED** [review]

File: core/src/avm2/e4x.rs (hunk 1 above)
Comment (sleepycatcoding) [reply core/src/avm2/e4x.rs:249]: Can cleanup in follow up PR.

File: core/src/avm2/object/xml_object.rs (hunk 2 above)
Comment (sleepycatcoding) [reply core/src/avm2/object/xml_object.rs:365]: Yeah, I also realized this myself, I should have left the `null` and `undefined` handling in AS3 `isXMLName` function. Can create in follow up PR.

**Pull request merged**